stl/inc/chrono

```diff
@@ -1480,6 +1481,33 @@ namespace chrono {
         return _Result;
     }
 
+    template <class _ToPeriod, class _FromPeriod>
+    constexpr bool _Duration_cast_would_underflow = _FromPeriod::den > INTMAX_MAX / _ToPeriod::num;

```

Comment:
This criterion incorrectly detects underflow. Example: https://godbolt.org/z/79zfq3qaT

```cpp
#include <chrono>
#include <cstdint>
#include <iostream>
#include <ratio>
using namespace std;
using namespace std::chrono;

int main() {
    using FromPeriod = femto;
    using ToPeriod = ratio<86399, femto::den>; // 86399 is prime

    if (FromPeriod::den > INTMAX_MAX / ToPeriod::num) {
        cout << "ToPeriod::num * FromPeriod::den would overflow." << endl;
    } else {
        cout << "ToPeriod::num * FromPeriod::den would NOT overflow." << endl;
    }

    using FromDuration = duration<long long, FromPeriod>;
    using ToDuration = duration<long long, ToPeriod>;

    cout << duration_cast<ToDuration>(FromDuration{86399 * 5}).count() << endl;
}
```

```
ToPeriod::num * FromPeriod::den would overflow.
5
```

This is because `duration_cast` is specified with `ratio_divide` which uses GCD logic, see WG21-N4910 [[time.duration.cast]/2](https://eel.is/c++draft/time.duration.cast#2).

In this example, multiplying `femto::den` by anything larger than `9223` will overflow 2<sup>63</sup> - 1. `86399` is larger than that threshold.

However, `duration_cast` does not underflow - it reports `5`, because we're exactly "coarsening" `86399 * 5` femtoseconds into `5` ticks of 86399 femtoseconds each. This is because it uses `ratio_­divide<FromPeriod, ToPeriod>`, which cancels out the `femto`s.

`_Ratio_divide_sfinae` would handle such cases: https://github.com/microsoft/STL/blob/7f04137880e1c3df5125c1baf808f16f399dee2e/stl/inc/ratio#L158-L160

Now, this may not be an *actual* bug because it appears that you control all of the To durations, and they are all of the form `ratio<ToNum, 1>` (so there is no to-denominator that could cancel things out). Ideally we would `static_assert` this, or at the very least comment it. (Uncommented machinery that makes special-case assumptions is dangerous because it can be copy-pasted or imitated in general contexts.)

Reply:
You're right, and actually I thought of when writing up the commit messages. It looks easy enough to fix this by using ratio_divide here (which is confusingly named, which is why I didn't use it in the first place.).

We can't use sfnae (or concepts) to directly detect the failure of _Ratio_divide_sfinae because we want to preserve the behavior for overflows in the numerator of the conversion (since in that case the conversion will truncate, perhaps well short of the correct result)

Reply:
Maybe we can use _Safe_mult directly though

Reply:
I like the idea of checking the overflow in only one place